### PR TITLE
Support Fiber#transfer-based fiber schedulers

### LINKED ISSUE #4003: Migrate dataloader to use Fiber#transfer for modern scheduler compatibility
**Is your feature request related to a problem? Please describe.**

The current implementation of Dataloader uses `yield` / `resume` for Fiber management which makes it incompatible with modern Ruby 3.1 based fiber schedulers which rely on `transfer` instead (see https://github.com/bruno-/fiber_scheduler_list for a reference).

AFAICT there is no production ready Ruby 3.1 based scheduler that can be used today with the GraphQL gem.

The documentation for non-blocking dataloading at https://graphql-ruby.org/dataloader/nonblocking.html mentions "not having been able to figure out how to make Async::Scheduler work". The reason is the same - `Async::Scheduler` does not work with `yield`/ `resume`.

**Describe the solution you'd like**

I've had some Slack interaction with Bruno Sutic, author of the above scheduler list as well as the [fiber_scheduler](https://github.com/bruno-/fiber_scheduler) gem.

According to him migrating toward using `transfer` would be the right way to go for the GraphQL gem as he expects that to be the default paradigm for Fiber schedulers going forward.

I'm not sure how relevant it is to keep backwards compatibility with ruby 3.0 based schedulers. If needed then I guess we would need two different fiber management implementations and offer developers a configuration option to switch between them.

**Describe alternatives you've considered**

I spent a few hours trying to implement the change myself. I got the [async_dataloader_spec.rb](https://github.com/rmosolgo/graphql-ruby/blob/master/spec/graphql/dataloader/async_dataloader_spec.rb) tests passing with ~~both `Async::Scheduler` and~~ `FiberScheduler` but a lot of deeper integration tests ended up failing and it appears too time consuming for me to make sense of all the parts to make progress with the time I have on hand.

### PR DESCRIPTION
**Pull request by rmosolgo**
This adds a mode to Dataloader which works with `Fiber#transfer`-based schedulers (eg, `fiber_scheduler`, `Async::Scheduler`(?) ). It requires `use GraphQL::Dataloader, nonblocking: true, fiber_control_mode: :transfer`

Fixes #4003

TODO: 

- [x] FiberScheduler fails on Ruby 3.2
- [x] Try adding `Async::Scheduler`, too

### CODE REVIEW

**Comment by ioquatix** [comment]
For old rubies, resume/yield and transfer cannot be mixed, but going forward, it's acceptable.

I'd suggest simply not supporting Rubies where it can't be mixed, i.e. depend on Async v2+ (Ruby 3.1+)

**Comment by rmosolgo** [comment]
That's good to keep in mind. For now, I'm just trying to figure out how to get the build not to segfault 😩 

I haven't exactly figured out what's making it go wild, but evidently it doesn't like to "`sleep 0.1`" from inside fibers...

**Comment by ioquatix** [comment]
If you want to chat or show me the issue, let me know. I'm happy to take a look.

**Comment by ioquatix** [comment]
What kind of segfault are you seeing?

**Comment by rmosolgo** [comment]
It runs fine locally, but in CI, `yield`ing from inside `Fiber.new(blocking: false) { ... }` crashes for at _least_ `Async::Scheduler` (https://github.com/rmosolgo/graphql-ruby/actions/runs/5730999250/job/15530962687) and `FiberScheduler` (https://github.com/rmosolgo/graphql-ruby/actions/runs/5732062739/job/15534264398), but not for DummyScheduler (plain-Ruby, copied from Ruby's tests). 

I just tried refactoring to pass an object instead of `yield`, but that didn't work either :S Maybe `Fiber.new(blocking: false).resume` isn't _supposed_ to work 🤔 ?

**Comment by ioquatix** [comment]
What version of Ruby was failing? Can you share the logs/errors?

**Comment by ioquatix** [comment]
Backticks and system should be handled identically.

**Comment by rmosolgo** [comment]
How about those failures on GitHub actions? They're Ruby 3.2.2, with the segfault message logged there. I'd be happy to add more logging if there's something that'd be helpful to you!

https://github.com/rmosolgo/graphql-ruby/actions/runs/5730999250/job/15530962687

**Comment by ioquatix** [comment]
Looks like it's dying on https://github.com/ruby/ruby/blob/v3_2_2/process.c#L4849

If you have spare time, is there a chance you can try to make a small repro? Otherwise, I'll take a look later this week.

File: spec/graphql/dataloader/async_dataloader_spec.rb [hunk 1]
```diff
@@ -296,13 +296,39 @@ def do_sleep(duration)
...
+            query(Query)
+            use GraphQL::Dataloader, nonblocking: true, fiber_control_mode: :transfer
+          end
+
+          start = Time.now
+          puts "Starting..."
+          Fiber.set_scheduler(Async::Scheduler.new)
+          pp MySchema.execute("{ v1: value v2: value }").to_h
+          puts "Ended"
+          pp [:duration, Time.now - start]
```
Comment (ioquatix) [inline spec/graphql/dataloader/async_dataloader_spec.rb:329]: You may want to add `Fiber.set_scheduler(nil)` to avoid leaking state.

**Comment by rmosolgo** [comment]
Here's a minimal reproduction: 

https://github.com/rmosolgo/graphql-ruby/blob/3bade9c11a2a08fb9e932c800936e20ceac5a546/spec/graphql/dataloader/async_dataloader_spec.rb#L308-L329

It works fine if I run it on my machine ... but fails on GH Actions :S

**Comment by ioquatix** [comment]
It probably means race condition.

**Comment by rmosolgo** [comment]
Thanks for the hint about `IO_EVENT_SELECTOR` over on the ruby bug tracker (sorry if I opened that in the wrong spot 🤦 )! The tests now _run_ -- they just fail. But that's OK -- I'll take a closer look soon but I think it's just because of the latency in that io-event backend.

**Comment by ioquatix** [comment]
It's fine, whatever gets attention :) I am glad that helped. I'll investigate further on my end.

**Comment by ioquatix** [comment]
The problem seems to be this code:

```ruby
          # These fibers were previously waiting for sources to load data,
          # resume them. (They might wait again, in which case, re-enqueue them.)
          resume(f)
          if f.alive?
            next_fibers << f
          end
        end
```

Specifically, `resume(f)`.

You are resuming a fiber which is waiting in `process_wait`. If you do that, the following code is triggered:

```clang
static
VALUE process_wait_transfer(VALUE _arguments) {
	struct process_wait_arguments *arguments = (struct process_wait_arguments *)_arguments;
	
	VALUE result = IO_Event_Selector_fiber_transfer(arguments->selector->backend.loop, 0, NULL);
	
	if (arguments->waiting->result) {
		return IO_Event_Selector_process_status_wait(arguments->pid);
	} else {
		return Qfalse;
	}
}
```

Because the event loop has not actually waited successfully (on the process finishing), `arguments->waiting->result == 0`, and so we return `Qfalse`. Because of that, CRuby itself dies:

```clang
        VALUE status = rb_process_status_wait(pid, 0); # this is Qfalse.
        struct rb_process_status *data = RTYPEDDATA_DATA(status);
```

If you schedule a fiber, you must not resume it. e.g. `fiber = Fiber.schedule{}` should not call `resume` arbitrarily, unless you have specifically called `Fiber.yield`. I'll see if I can make this into an error, so that it's more obvious what the correct behaviour is and it fails with an exception rather than a segfault.

**Comment by ioquatix** [comment]
By the way, if you want to wait on a `Fiber.schedule` to complete, you can basically do this:

```ruby
completions = Thread::Queue.new

work.each do |job|
  Fiber.schedule do
    completions << job.call
  end
end
```

You can count the number of jobs and completions or use some other mechanism to know when all jobs are completed.

**Comment by rmosolgo** [comment]
Ok, I think that makes sense, thanks... Previously, the code could assume that if the main fiber was running, then the other fiber(s) had terminated or yielded.  But that assumption was wrong (even _then_, and now it's raising an error) because in fact, the other fiber could be waiting on IO and so the main fiber could have been given back control. 

Completion is one possibility for a stopped fiber -- thanks for the suggestion on handling that better. 

The other possibility is an explicit transfer of control (via `yield` or a call to `.transfer` on the parent fiber): 

https://github.com/rmosolgo/graphql-ruby/blob/e48f302a58f0eaf946737e007a47ec590be42c17/lib/graphql/dataloader.rb#L92-L104

Somehow I'll have to distinguish between those to determine whether the fiber should be re-enqueued to be restarted.

**Comment by rmosolgo** [comment]
Working this out in #4625 and #4727 instead.

**Pull request closed without merging**